`v2.26.0` changes cause build failures in some downstream providers

### SDK version
<!---
Inspect your go.mod as below to find the version, and paste the result between the ``` marks below.

go list -m github.com/hashicorp/terraform-plugin-sdk/...

If you are not running the latest version of the SDK, please try upgrading
because your bug may have already been fixed.

If the command above doesn't yield any results, it means you may either be using v1 of the SDK or
have not have migrated to the standalone SDK yet. See https://www.terraform.io/docs/extend/plugin-sdk.html
for more.
-->

[`v2.26.0`](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/CHANGELOG.md#2260-march-20-2023)

### Actual Behavior
<!--
What actually happened?
-->

Changes made in `v2.26.0` cause [build failures in some downstream providers](https://github.com/hashicorp/terraform-provider-aws/actions/runs/4476337082/jobs/7866604693?pr=30145):

```
go build -o terraform-plugin-dir/registry.terraform.io/hashicorp/aws/99.99.99/$(go env GOOS)_$(go env GOARCH)/terraform-provider-aws .
...
# github.com/hashicorp/terraform-provider-aws/internal/tfresource
Error: internal/tfresource/errors.go:22:20: impossible type assertion: err.(*resource.TimeoutError)
	*resource.TimeoutError does not implement error (missing Error method)
Error: internal/tfresource/errors.go:30:7: impossible type switch case: *resource.TimeoutError
	err (variable of type error) cannot have dynamic type *resource.TimeoutError (missing Error method)
Error: internal/tfresource/errors.go:35:7: impossible type switch case: *resource.UnexpectedStateError
	err (variable of type error) cannot have dynamic type *resource.UnexpectedStateError (missing Error method)
Error: internal/tfresource/retry.go:26:29: cannot use func() *resource.RetryError {…} (value of type func() *resource.RetryError) as type resource.RetryFunc in argument to Retry
Error: internal/tfresource/retry.go:237:18: c.WaitForStateContext undefined (type *resource.StateChangeConf has no field or method WaitForStateContext)
Error: internal/tfresource/wait.go:53:22: stateConf.WaitForStateContext undefined (type *resource.StateChangeConf has no field or method WaitForStateContext)
# github.com/hashicorp/terraform-provider-aws/internal/service/athena
Error: internal/service/athena/database.go:275:15: cannot use queryExecutionStateRefreshFunc(ctx, conn, qeid) (value of type "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".StateRefreshFunc) as type retry.StateRefreshFunc in struct literal
Error: internal/service/athena/database.go:280:31: executionStateConf.WaitForStateContext undefined (type *"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".StateChangeConf has no field or method WaitForStateContext)
# github.com/hashicorp/terraform-provider-aws/internal/service/glacier
Error: internal/service/glacier/vault_lock.go:209:12: cannot use vaultLockRefreshFunc(ctx, conn, vaultName) (value of type "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".StateRefreshFunc) as type retry.StateRefreshFunc in struct literal
Error: internal/service/glacier/vault_lock.go:214:22: stateConf.WaitForStateContext undefined (type *"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".StateChangeConf has no field or method WaitForStateContext)
...
```

--- a/.changes/unreleased/BUG FIXES-20230321-085655.yaml
+++ b/.changes/unreleased/BUG FIXES-20230321-085655.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: 'helper/resource: Prevented build errors with type aliasing added in v2.26.0'
+time: 2023-03-21T08:56:55.662651-04:00
+custom:
+  Issue: "1176"

--- a/helper/resource/aliases.go
+++ b/helper/resource/aliases.go
@@ -92,7 +92,7 @@ type RetryFunc = retry.RetryFunc
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
 func RetryContext(ctx context.Context, timeout time.Duration, f RetryFunc) error {
-	return retry.RetryContext(ctx, timeout, retry.RetryFunc(f))
+	return retry.RetryContext(ctx, timeout, f)
 }
 
 // Retry is a basic wrapper around StateChangeConf that will just retry
@@ -101,7 +101,7 @@ func RetryContext(ctx context.Context, timeout time.Duration, f RetryFunc) error
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
 func Retry(timeout time.Duration, f RetryFunc) error {
-	return retry.Retry(timeout, retry.RetryFunc(f))
+	return retry.Retry(timeout, f)
 }
 
 // RetryError is the required return type of RetryFunc. It forces client code

--- a/helper/resource/aliases.go
+++ b/helper/resource/aliases.go
@@ -42,19 +42,19 @@ func PrefixedUniqueId(prefix string) string {
 
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type NotFoundError retry.NotFoundError
+type NotFoundError = retry.NotFoundError
 
 // UnexpectedStateError is returned when Refresh returns a state that's neither in Target nor Pending
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type UnexpectedStateError retry.UnexpectedStateError
+type UnexpectedStateError = retry.UnexpectedStateError
 
 // TimeoutError is returned when WaitForState times out
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type TimeoutError retry.TimeoutError
+type TimeoutError = retry.TimeoutError
 
 // StateRefreshFunc is a function type used for StateChangeConf that is
 // responsible for refreshing the item being watched for a state change.
@@ -69,19 +69,19 @@ type TimeoutError retry.TimeoutError
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type StateRefreshFunc retry.StateRefreshFunc
+type StateRefreshFunc = retry.StateRefreshFunc
 
 // StateChangeConf is the configuration struct used for `WaitForState`.
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type StateChangeConf retry.StateChangeConf
+type StateChangeConf = retry.StateChangeConf
 
 // RetryFunc is the function retried until it succeeds.
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type RetryFunc retry.RetryFunc
+type RetryFunc = retry.RetryFunc
 
 // RetryContext is a basic wrapper around StateChangeConf that will just retry
 // a function until it no longer returns an error.
@@ -109,7 +109,7 @@ func Retry(timeout time.Duration, f RetryFunc) error {
 //
 // Deprecated: Use helper/retry package instead. This is required for migrating acceptance
 // testing to terraform-plugin-testing.
-type RetryError retry.RetryError
+type RetryError = retry.RetryError
 
 // RetryableError is a helper to create a RetryError that's retryable from a
 // given error. To prevent logic errors, will return an error when passed a

